src/core/fetchers/segment/prioritizer.ts

```diff
@@ -136,87 +223,236 @@ export default class ObservablePrioritizer<T> {
+   * The task will be removed from the waiting queue in the process.
+   * @param {number} index
+   */
+  private _startTask(index : number) : void {
+    const task = this._waitingQueue.splice(index, 1)[0];
+    task.trigger.next(true);
+  }
+
+  /**
+   * Move back pending task - which is at the given index in the _pendingTasks
+   * array - back to the waiting queue and interrupt it.
+   * @param {number} index
+   */
+  private _interruptPendingTask(index : number) : void {
+    // Stop task and put it back in the waiting queue
+    const task = this._pendingTasks.splice(index, 1)[0];
+    this._waitingQueue.push(task);
+    if (this._pendingTasks.length === 0) {
+      this._minPendingPriority = null;
+    } else if (this._minPendingPriority === task.priority) {
+      this._minPendingPriority = Math.min(...this._pendingTasks.map(t => t.priority));
+    }
+    task.trigger.next(false);
+  }
+
+  /**
+   * Logic ran when a task begin.
+   * @param {Object} task
+   */
+  private _onTaskBegin(task : IPrioritizerTask<T>) : void {
+    this._minPendingPriority = this._minPendingPriority === null ?
+      task.priority :
+      Math.min(this._minPendingPriority, task.priority);
+    this._pendingTasks.push(task);
+  }
+
+  /**
+   * Logic ran when a task has ended.
+   * @param {Object} task
+   */
+  private _onTaskEnd(task : IPrioritizerTask<T>) : void {
+    const pendingTasksIndex = arrayFindIndex(this._pendingTasks,
+                                             (elt) => elt.observable === task.observable);
+    if (pendingTasksIndex < 0) {
+      return; // Happen for example when the task has been interrupted
+    }
+
+    task.trigger.complete();
+
+    this._pendingTasks.splice(pendingTasksIndex, 1);
+
+    if (this._pendingTasks.length > 0) {
+      return; // still waiting for Observables to finish
+    }
+
+    this._minPendingPriority = null;
+
+    if (this._waitingQueue.length === 0) {
+      return; // no more task to do
+    }
+
+    // Calculate minimum waiting priority
+    this._minPendingPriority = this._waitingQueue.reduce((acc : number | null, elt) => {
+      return acc === null || acc > elt.priority ? elt.priority :
+                                                  acc;
+    }, null);
+
+    // Start all tasks with that minimum priority
+    for (let i = 0; i < this._waitingQueue.length; i++) {
+      const elt = this._waitingQueue[i];
+      if (elt.priority === this._minPendingPriority) {
+        this._startTask(i);
+        i--; // previous operation should have removed that element from the
+             // the waiting queue
       }
-    };
+    }
+  }
 
-    this._numberOfPendingObservables++;
-    return obs
-             .pipe(finalize(onObservableFinish));
+  /**
+   * Return `true` if the given task should be started immediately.
+   * @param {Object} task
+   * @returns {boolean}
+   */
+  private _hasPriority(task : IPrioritizerTask<T>) : boolean {
```

Comment:
Strange function name IMO. I first though this checked if a task has an assigned priority, which seems to always be the case.

Reply:
Ok. Maybe `_shouldBeStartedNow`?

Reply:
Yes

Reply:
Done except I called it `_canBeStartedNow` in the end